### ISSUE TITLE
No executable process is found after transformation

### ISSUE THREAD

**Issue by nicpuppa** [camunda/camunda#19834]
**Describe the bug**

This error was reported by our internal error reporting tool. 

It seems that an executable process is not found either in cache or in memory.

**To Reproduce**

Not clear

**Expected behavior**

Not clear

**Log/Stacktrace**

Link to [logs](https://console.cloud.google.com/logs/query;query=error_groups.id%3D%22CMqfmayEhZT_4QE%22%0AlogName:%22stdout%22%0Aresource.type%3D%22k8s_container%22%0Aresource.labels.namespace_name%3D%221868e9b6-b6f1-46bc-81f5-4e15451e41b2-zeebe%22%0Aresource.labels.location%3D%22northamerica-northeast2%22%0Aresource.labels.cluster_name%3D%22worker-5%22%0Aresource.labels.project_id%3D%22camunda-saas-prod%22%0Aresource.labels.container_name%3D%22zeebe%22%0Aresource.labels.pod_name%3D%22zeebe-2%22;cursorTimestamp=2024-06-28T08:41:25.022869857Z;startTime=2024-06-28T08:11:55.022Z;endTime=2024-06-28T09:11:55.022Z?project=camunda-saas-prod)
Download [logs](https://drive.google.com/drive/folders/1F57PeVkmY0nKC5XHySI3bzGkUHaqolt5)

<details><summary>Full Stacktrace</summary>
 <p>

```
java.util.NoSuchElementException: Expected to find executable process in persisted process with key '2251799813685409', but after transformation no such executable process could be found.
	at io.camunda.zeebe.engine.state.deployment.DbProcessState.lambda$updateInMemoryState$1(DbProcessState.java:244) ~[zeebe-workflow-engine-8.5.3.jar:8.5.3]
	at java.base/java.util.Optional.orElseThrow(Unknown Source) ~[?:?]
	at io.camunda.zeebe.engine.state.deployment.DbProcessState.updateInMemoryState(DbProcessState.java:241) ~[zeebe-workflow-engine-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.engine.state.deployment.DbProcessState.lookupProcessByIdAndPersistedVersion(DbProcessState.java:400) ~[zeebe-workflow-engine-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.engine.state.deployment.DbProcessState.getLatestProcessVersionByProcessId(DbProcessState.java:292) ~[zeebe-workflow-engine-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.engine.processing.deployment.StartEventSubscriptionManager.isLatestProcess(StartEventSubscriptionManager.java:67) ~[zeebe-workflow-engine-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.engine.processing.deployment.StartEventSubscriptionManager.tryReOpenStartEventSubscription(StartEventSubscriptionManager.java:58) ~[zeebe-workflow-engine-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.engine.processing.deployment.DeploymentCreateProcessor.processDistributedCommand(DeploymentCreateProcessor.java:114) ~[zeebe-workflow-engine-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.engine.processing.streamprocessor.DistributedTypedRecordProcessor.processRecord(DistributedTypedRecordProcessor.java:31) ~[zeebe-workflow-engine-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.engine.Engine.process(Engine.java:127) ~[zeebe-workflow-engine-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.batchProcessing(ProcessingStateMachine.java:369) ~[zeebe-stream-platform-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.lambda$processCommand$3(ProcessingStateMachine.java:282) ~[zeebe-stream-platform-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:116) ~[zeebe-db-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.processCommand(ProcessingStateMachine.java:282) ~[zeebe-stream-platform-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.stream.impl.ProcessingStateMachine.tryToReadNextRecord(ProcessingStateMachine.java:240) ~[zeebe-stream-platform-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) [zeebe-scheduler-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:47) [zeebe-scheduler-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:122) [zeebe-scheduler-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:130) [zeebe-scheduler-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:108) [zeebe-scheduler-8.5.3.jar:8.5.3]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:227) [zeebe-scheduler-8.5.3.jar:8.5.3]
```

</p>
</details>

**Environment:**
- Zeebe Version: 8.5.3

**Comment by nicpuppa** [comment]
I'm going to download data from the cluster

**Comment by korthout** [comment]
Triage:
- seems scary, and should be investigated soon
- we'll pull it into the current iteration

**Comment by remcowesterhoud** [comment]
Small update from what I found so far:

- This can only happen when there's multiple BPMN processes in a single deployment
- The affected cluster seems to be recovered after a broker restart
- It appears the `DbProcessState` tries to get a process by id and version out of the column family. It finds the correct process, however the resource inside of this is wrong. The resource belongs to the other process in the deployment.

I'll continue to dig deeper and see if I can find out where it gets the wrong resource. At this time I'm not too worried about the impact as the deployment results in an exception (no wrong data in the state) and it eventually recovered.

**Comment by remcowesterhoud** [comment]
I just noticed something interesting. Looking at the deployment create command we can see that there is 2 processes deployed at once, both with the **same resource name**. As a result we write 2 `Process.CREATED` events for each of these, since we do a match on resource name:

```java
private void createBpmnResources(final DeploymentRecord deploymentEvent) {
    deploymentEvent.processesMetadata().stream()
        .filter(not(ProcessMetadata::isDuplicate))
        .forEach(
            metadata -> {
              for (final DeploymentResource resource : deploymentEvent.getResources()) {
                if (resource.getResourceName().equals(metadata.getResourceName())) {
                  stateWriter.appendFollowUpEvent(
                      metadata.getKey(),
                      ProcessIntent.CREATED,
                      new ProcessRecord().wrap(metadata, resource.getResource()));
                }
              }
            });
  }
```

I reckon this is where the bug lives. We take the metadata from process 1, but the resource of process 2. Both these resources have a **different process id**! Later on we run into an exception here:

```java
    final ExecutableProcess executableProcess =
        definitions.stream()
            .filter(
                process -> BufferUtil.equals(persistedProcess.getBpmnProcessId(), process.getId()))
            .findFirst()
            .orElseThrow(
                () ->
                    new NoSuchElementException(
                        String.format(
                            "Expected to find executable process in persisted process with key '%s',"
                                + " but after transformation no such executable process could be found.",
                            persistedProcess.getKey())));
```

The filter is the cause of this exception. We try to match the id of the persisted process with the id of the process. These no longer match due to the bug above.


--------------

**TL;DR** 

- Bug occurs when there is a deployment containing multiple BPMN resources
    - Both resources must have the same resource name (filename)
    - Both resources must have a different process id
- This only happens on the distributed `Deployment.CREATE` command.
    - Bug occurs in a rejection of the Command, no invalid data is persisted
